File: ios/uniMag-SDK/UMCardData.m

#import "UMCardData.h"
// idtech 714-761-6368
//------------------------------------------------------------------------------
#pragma mark - Utilities
//------------------------------------------------------------------------------

static BOOL isBitSet(Byte byte, int bitIndex) {
  Byte mask = (1 << bitIndex);
  return byte & mask;
}

//------------------------------------------------------------------------------
#pragma mark - Initialize
//------------------------------------------------------------------------------

@implementation UMCardData

@synthesize byteData = _byteData;

@synthesize isValid = _isValid;
@synthesize isEncrypted = _isEncrypted;
@synthesize isAesEncrypted = _isAesEncrypted;

@synthesize track1 = _track1;
@synthesize track2 = _track2;
@synthesize track3 = _track3;
@synthesize track1_encrypted = _track1_encrypted;
@synthesize track2_encrypted = _track2_encrypted;
@synthesize track3_encrypted = _track3_encrypted;

@synthesize serialNumber = _serialNumber;
@synthesize KSN = _KSN;

/*
- (void)dealloc {
    [_byteData   release];

    [_track1 release];
    [_track2 release];
    [_track3 release];
    [_track1_encrypted release];
    [_track2_encrypted release];
    [_track3_encrypted release];

    [_serialNumber release];
    [_KSN release];

    [super dealloc];
}
*/

- (id)initWithBytes:(NSData *)cardData {
  self = [super init];
  if (self && cardData) {
    self->_byteData = [cardData copy];

    const UInt8 *bytes = cardData.bytes;
    const NSUInteger len = cardData.length;

    // check validity and parse
    if (len >= 1) {
      // is valid encrypted swipe?
      if (bytes[0] == '\x02') {
        self->_isEncrypted = TRUE;
        [self verifyAndParse_encrypted:cardData];
      }
      // is valid unencrypted swipe?
      else {
        self->_isEncrypted = FALSE;
        [self verifyAndParse_unencrypted:cardData];
      }
    }
  }
  return self;
}

- (void)verifyAndParse_encrypted:(NSData *)cardData {
  /*
    Encrypted swipe format: STX LenL LenH <payload> CheckXor CheckSum ETX
    The rest... there's a document.
  */

  // things to fill out
  NSData *tracks[3] = {nil, nil, nil};
  NSData *tracks_enc[3] = {nil, nil, nil};
  NSData *serialNumber = nil;
  NSData *ksn = nil;
  NSData *payload = nil;
  BOOL isAES = FALSE; // true if AES, false if TDES
  BOOL isSwipeDataValid = FALSE;

  // shorthand
  const UInt8 *bytes = cardData.bytes;
  const NSUInteger len = cardData.length;

  // Verify
  if (len < 6)
    return;
  // STX ETX
  if (bytes[0] != '\x02' || bytes[len - 1] != '\x03')
    return;
  // Length
  int payloadLen = (bytes[2] << 8) + (bytes[1]);
  if (payloadLen + 6 != len)
    return;
  // CheckXor and CheckSum
  UInt8 cksum = 0, ckxor = 0;
  for (int i = 3; i < len - 3; i++) {
    ckxor ^= bytes[i];
    cksum += bytes[i];
  }
  if (bytes[len - 2] != cksum || bytes[len - 3] != ckxor)
    return;

  // parse
  {
#define CHECK_INDEX(I) \
  {                    \
    if ((I) > len - 3) \
      goto stopParse;  \
  }
    int idx = 0;

    // get track1, track2, track3 length
    int trackLens[3];
    idx = 5;
    CHECK_INDEX(idx + 3);
    for (int i = 0; i < 3; i++)
      trackLens[i] = bytes[idx + i];

    // get masked track
    int trackLensSum = 0;
    idx = 10;
    for (int i = 0; i < 3; i++) {
      // skip if len is 0 or presence flag not set
      if (trackLens[i] == 0 || !isBitSet(bytes[8], i))
        continue;
      CHECK_INDEX(idx + trackLens[i]);
      tracks[i] = [cardData subdataWithRange:NSMakeRange(idx, trackLens[i])];
      idx += trackLens[i];
      trackLensSum += trackLens[i];
    }

    // determine encryption type (TDES or AES)
    idx = 8;
    CHECK_INDEX(idx + 1);
    Byte encType = ((bytes[idx] >> 4) & 0x03);
    if (encType == 0x00)
      isAES = FALSE;
    else if (encType == 0x01)
      isAES = TRUE;
    else
      goto stopParse;

    // TODO: none encrypting encrypted swipe

    // get whole payload for merchant processing

    // get encrypted section
    int trackLens_enc[3];
    int encryptionBlockSize = (isAES ? 16 : 8);
    for (int i = 0; i < 3; i++) {
      trackLens_enc[i] = ceil(trackLens[i] / (double)encryptionBlockSize) *
                         encryptionBlockSize;
    }
    int trackLensSum_enc = 0;
    idx = 10 + trackLensSum;
    int payloadbegin = idx; //-4 according to program from cayan
    int payloadend = payloadbegin;
    for (int i = 0; i < 3; i++) {
      // skip if len is 0 or presence flag not set
      if (trackLens_enc[i] == 0 || !isBitSet(bytes[9], i))
        continue;
      CHECK_INDEX(idx + trackLens_enc[i]); // verify this won't overflow
      tracks_enc[i] =
          [cardData subdataWithRange:NSMakeRange(idx, trackLens_enc[i])];
      idx += trackLens_enc[i];
      trackLensSum_enc += trackLens_enc[i];
      payloadend = idx;
    }
    payload = [cardData
        subdataWithRange:NSMakeRange(payloadbegin, payloadend - payloadbegin)];

    // get KSN
    if ((isBitSet(bytes[9], 7)) &&
        (isBitSet(bytes[9], 0) || isBitSet(bytes[9], 1) ||
         isBitSet(bytes[9], 2))) {
      idx = (int)(len - 3 - 10);
      if (idx < 10 + trackLensSum + trackLensSum_enc)
        goto stopParse;
      ksn = [cardData subdataWithRange:NSMakeRange(idx, 10)];
    }

    // get serial number
    if (isBitSet(bytes[8], 7)) {
      idx = (int)len - 3 - 10 - (ksn ? 10 : 0);
      if (idx < 10 + trackLensSum + trackLensSum_enc)
        goto stopParse;
      serialNumber = [cardData subdataWithRange:NSMakeRange(idx, 10)];
    }

    // all checks and parsing succeeded
    isSwipeDataValid = TRUE;

#undef CHECK_INDEX
  }
stopParse:

  // save result
  self->_isValid = isSwipeDataValid;
  self->_isAesEncrypted = isAES;
  self->_payload = payload;
  self->_track1 = tracks[0];
  self->_track2 = tracks[1];
  self->_track3 = tracks[2];
  self->_track1_encrypted = tracks_enc[0];
  self->_track2_encrypted = tracks_enc[1];
  self->_track3_encrypted = tracks_enc[2];
  self->_serialNumber = serialNumber;
  self->_KSN = ksn;
}

- (void)verifyAndParse_unencrypted:(NSData *)cardData {
  /*
    Unencrypted swipe is the concatenation of one or more tracks (as listed
    below), then ended by a "\x0D". Track formats (in regular expression):
    ISO_1: "[\x25][^\x3F]+[\x3F]"
    ISO_2: "[\x3B][^\x3F]+[\x3F]"
    JIS  : "[\x7F][^\x7F]+[\x7F]"
    */

  // things to fill out
  NSData *tracks[3] = {nil, nil, nil};
  NSData *payload = nil;
  BOOL isSwipeDataValid = FALSE;

  // shorthand
  const UInt8 *bytes = cardData.bytes;
  const NSUInteger len = cardData.length;

  // parse manually
  BOOL ps_isOutsideTrack =
      TRUE; // parser state: false if outside of track, true if inside
  BOOL ps_isISO; // ISO or JIS track
  int ps_trackStart = 0; // starting point of this track
  int ps_tracksI = 0; // track index
  for (int i = 0; i < len; i++) {
    const Byte b = bytes[i];
    if (ps_isOutsideTrack) {
      if (b == '\x25' || b == '\x3B') {
        ps_isOutsideTrack = FALSE;
        ps_isISO = TRUE;
      } else if (b == '\x7F') {
        ps_isOutsideTrack = FALSE;
        ps_isISO = FALSE;
      } else if (b == '\x0D') {
        // if it is last char, parsing's done
        // and the whole thing is valid.
        if (i == len - 1)
          isSwipeDataValid = TRUE;
        // there's unexpected char after ending character \x0D
        else
          break;
      }
      // unexpected char when expecting a track
      else
        break;
    } else {
      // skip track content
      if (ps_isISO ? (b != '\x3F') : (b != '\x7F'))
        continue;
      // track ended
      else {
        // save track
        tracks[ps_tracksI++] = [NSData dataWithBytes:bytes + ps_trackStart
                                              length:i + 1 - ps_trackStart];
        ps_trackStart = i + 1;
        ps_isOutsideTrack = TRUE;
      }
    }
  }
  payload = [cardData subdataWithRange:NSMakeRange(0, len)];

  // save result
  self->_isValid = isSwipeDataValid;
  self->_track1 = tracks[0];
  self->_track2 = tracks[1];
  self->_track3 = tracks[2];
}

@end

File: android/src/main/java/com/oncethere/idtechmsraudio/UmCardData.java
package com.oncethere.idtechmsraudio;

/**
 * Created by aamirshehzad on 6/28/15.
 */
public class UmCardData {
  public byte[] byteData; //Raw data which was used to initialize this object.

  //Meta
  public boolean isValid; //All available fields were successfully parsed from the raw data
  public boolean isEncrypted; //Whether the output is from an encrypting reader or a non-encrypting reader
  public boolean isAesEncrypted; //Only valid if isEncrypted==TRUE. If true, AES cipher is used, otherwise it's TDES

  //Parsed parts. Even if isValid==TRUE, some fields may be nil if the reader did not output them.
  // For example, reader may not read all tracks of a card.

  //  Tracks
  //    If non-encrypting reader is used, it is the full plain text card data
  //    If     encrypting reader is used, it is the partially masked plain text card data
  public byte[] track1;
  public byte[] track2;
  public byte[] track3;

  //  Encrypted Tracks. Only in encrypting reader output
  public byte[] track1_encrypted;
  public byte[] track2_encrypted;
  public byte[] track3_encrypted;

  //  Other. Only in encrypting reader output
  public byte[] serialNumber;
  public byte[] KSN;

  public UmCardData(byte[] cardData) {
    if (cardData != null) {
      this.byteData = cardData;
      final byte[] bytes = cardData;
      final Integer len = cardData.length;

      //check validity and parse
      if (len >= 1) {
        //is valid encrypted swipe?
        if (bytes[0] == 0x02) {
          this.isEncrypted = true;
          this.verifyAndParse_encrypted(cardData);
        }
        //is valid unencrypted swipe?
        else {
          this.isEncrypted = false;
          this.verifyAndParse_unencrypted(cardData);
        }
      }
    }
  }

  private static boolean isBitSet(byte abyte, int bitIndex) {
    byte b = 1;
    byte mask = (byte)(b << (byte)bitIndex);
    byte result = (byte)(abyte & mask);
    if (result > 0)
      return true;
    return false;
  }

  private void verifyAndParse_encrypted(byte[] cardData) {
    //things to fill out
    ///NSData tracks [3] = {null, null, null};
    /// NSData track_enc[3] = {null, null, null};

    byte[] track1 = null;
    byte[] track2 = null;
    byte[] track3 = null;

    byte[] track_enc1 = null;
    byte[] track_enc2 = null;
    byte[] track_enc3 = null;

    byte[] serialNumber = null;
    byte[] ksn = null;
    boolean isAES = false; //true if AES, false if TDES
    boolean isSwipeDataValid = false;

    //shorthand
    final byte[] bytes = cardData;
    final Integer len = cardData.length;

    //Verify
    if (len < 6)
      return;
    // STX ETX
    if (bytes[0] != 0x02 || bytes[len - 1] != 0x03)
      return;
    // Length
    int payloadLen = (bytes[2] << 8) + (bytes[1]);
    if (payloadLen + 6 != len)
      return;
    // CheckXor and CheckSum
    int cksum = 0, ckxor = 0;
    for (int i = 3; i < len - 3; i++) {
      ckxor ^= bytes[i];
      cksum += bytes[i];
    }
    if (bytes[len - 2] != cksum || bytes[len - 3] != ckxor)
      return;

    int idx = 0;

    //get track1, track2, track3 length
    int[] trackLens = new int[3];
    idx = 5;

    if (CHECK_INDEX(idx + 3, len) == false) {
      this.stopParse(isSwipeDataValid, isAES, track1, track2, track3, track_enc1, track_enc2, track_enc3, ksn, serialNumber);
      return;
    }
    for (int i = 0; i < 3; i++) {
      trackLens[i] = bytes[idx + i];
    }
    //get masked track
    int trackLensSum = 0;
    idx = 10;
    for (int i = 0; i < 3; i++) {
      //skip if len is 0 or presence flag not set
      if (trackLens[i] == 0 || !isBitSet(bytes[8], i))
        continue;

      if (CHECK_INDEX(idx + trackLens[i], len) == false) {
        this.stopParse(isSwipeDataValid, isAES, track1, track2, track3, track_enc1, track_enc2, track_enc3, ksn, serialNumber);
        return;
      }
      // tracks[i] = [cardData subdataWithRange: NSMakeRange(idx, trackLens[i])];
      switch (i) {
      case 0: {
        for (int j = idx, k = 0; j <= trackLens[i]; k++, j++) {
          track1[k] = cardData[j];
        }
        break;
      }
      case 1: {
        for (int j = idx, k = 0; j <= trackLens[i]; k++, j++) {
          track2[k] = cardData[j];
        }
        break;
      }
      case 2: {
        for (int j = idx, k = 0; j <= trackLens[i]; k++, j++) {
          track3[k] = cardData[j];
        }
        break;
      }
      }

      idx += trackLens[i];
      trackLensSum += trackLens[i];
    }

    //determine encryption type (TDES or AES)
    idx = 8;
    if (CHECK_INDEX(idx + 1, len) == false) {
      this.stopParse(isSwipeDataValid, isAES, track1, track2, track3, track_enc1, track_enc2, track_enc3, ksn, serialNumber);
      return;
    }
    byte encType = (byte)((bytes[idx] >> 4) & 0x03);
    if (encType == 0x00) {
      isAES = false;
    }
    //if the lowest significant bit of the high nibble is high, the encrypted track output is AES
    else if (encType == 0x01) {
      isAES = true;
    } else {
      this.stopParse(isSwipeDataValid, isAES, track1, track2, track3, track_enc1, track_enc2, track_enc3, ksn, serialNumber);
      return;
    }

    //TODO: none encrypting encrypted swipe

    //get encrypted section
    int[] trackLens_enc = new int[3];
    int encryptionBlockSize = (isAES ? 16 : 8);
    for (int i = 0; i < 3; i++) {

      trackLens_enc[i] = (int)Math.ceil(trackLens[i] / (double)encryptionBlockSize) * encryptionBlockSize;
    }
    int trackLensSum_enc = 0;
    idx = 10 + trackLensSum;
    for (int i = 0; i < 3; i++) {
      //skip if len is 0 or presence flag not set
      if (trackLens_enc[i] == 0 || !isBitSet(bytes[9], i))
        continue;

      if (CHECK_INDEX(idx + trackLens_enc[i], len) == false) {
        this.stopParse(isSwipeDataValid, isAES, track1, track2, track3, track_enc1, track_enc2, track_enc3, ksn, serialNumber);
        return;
      }
      switch (i) {
      case 0: {
        for (int j = idx, k = 0; j <= trackLens_enc[i]; k++, j++) {
          track_enc1[k] = cardData[j];
        }
        break;
      }
      case 1: {
        for (int j = idx, k = 0; j <= trackLens_enc[i]; k++, j++) {
          track_enc2[k] = cardData[j];
        }
        break;
      }
      case 2: {
        for (int j = idx, k = 0; j <= trackLens_enc[i]; k++, j++) {
          track_enc3[k] = cardData[j];
        }
        break;
      }
      }
      ///track_enc[i] = [cardData subdataWithRange: NSMakeRange(idx, trackLens_enc[i])];
      idx += trackLens_enc[i];
      trackLensSum_enc += trackLens_enc[i];
    }

    //get KSN /it looks as if you are checking for the high bit of the status byte (correct for KSN) and at least one of the individual track status bytes here.  Please consider not combining these checks as commented out below
    if ((isBitSet(bytes[9], 7)))
    //&& (isBitSet(bytes[9], 0)|| isBitSet(bytes[9], 1)|| isBitSet(bytes[9], 2)   )   )
    {
      idx = (int)(len - 3 - 10);
      //would [idx = (int)(len - 13);] work? If counting from ETX char, this logic follows to me
      if (idx < 10 + trackLensSum + trackLensSum_enc) {
        this.stopParse(isSwipeDataValid, isAES, track1, track2, track3, track_enc1, track_enc2, track_enc3, ksn, serialNumber);
        return;
      }

      for (int i = idx, j = 0; i <= 10; i++, j++) {
        ksn[j] = cardData[i];
      }

      ///[cardData subdataWithRange: NSMakeRange(idx, 10)];
    }

    //get serial number
    if (isBitSet(bytes[8], 7)) {
      idx = (int)len - 3 - 10 - (ksn != null ? 10 : 0);
      // as above, if counting from back (ETX) of data block, after confirming status byte value, length will be len - 23
      if (idx < 10 + trackLensSum + trackLensSum_enc) {
        this.stopParse(isSwipeDataValid, isAES, track1, track2, track3, track_enc1, track_enc2, track_enc3, ksn, serialNumber);
        return;
      }
      for (int i = idx, j = 0; i <= 10; i++, j++) {
        serialNumber[j] = cardData[i];
      }
      //// serialNumber = [cardData subdataWithRange: NSMakeRange(idx, 10)];
    }

    //all checks and parsing succeeded
    isSwipeDataValid = true;

    this.stopParse(isSwipeDataValid, isAES, track1, track2, track3, track_enc1, track_enc2, track_enc3, ksn, serialNumber);
  }

  private boolean CHECK_INDEX(int I, int len) {
    if ((I) > len - 3)
      return false; //goto stopParse;
    return true;
  }

  private void stopParse(boolean isSwipeDataValid, boolean isAES, byte[] track1, byte[] track2, byte[] track3, byte[] track_enc1, byte[] track_enc2, byte[] track_enc3, byte[] ksn, byte[] serialNumber) {
    this.isValid = isSwipeDataValid;
    this.isAesEncrypted = isAES;
    this.track1 = track1;
    this.track2 = track2;
    this.track3 = track3;
    this.track1_encrypted = track_enc1;
    this.track2_encrypted = track_enc2;
    this.track3_encrypted = track_enc3;
    this.serialNumber = serialNumber;
    this.KSN = ksn;
  }

  private void verifyAndParse_unencrypted(byte[] cardData) {
    //things to fill out
    /// NSData *tracks[3] = {nil, nil, nil};
    byte[] track1 = null;
    byte[] track2 = null;
    byte[] track3 = null;
    boolean isSwipeDataValid = false;

    //shorthand
    final byte[] bytes = cardData;
    final Integer len = cardData.length;

    //parse manually
    boolean ps_isOutsideTrack = true; //parser state: false if outside of track, true if inside
    boolean ps_isISO = false; //ISO or JIS track
    int ps_trackStart = 0; //starting point of this track
    int ps_tracksI = 0; //track index
    for (int i = 0; i < len; i++) {
      final Byte b = bytes[i];
      if (ps_isOutsideTrack) {
        if (b == 0x25 || b == 0x3B) //Please consider also referencing the LRC byte which in masked data is an * such as
        // byte b2 = bytes[i+1]
        //if (b == 0x3F)
        //if (b2 == 0x2A) {}
        {
          ps_isOutsideTrack = false;
          ps_isISO = true; //Please note that a start sentinel of a "%" is only an ID Tech start sentinel (reader added) and is not related to ISO.
              //In masked data, the format code, which begins the card data, is masked.  You may be able to get a good indication of card type from the first four digits of card data, but it would have limitations.
        } else if (b == 0x7F) {
          ps_isOutsideTrack = false;
          ps_isISO = false;
        } else if (b == 0x0D) //a carriage return is not expected for a reader such as the Shuttle//see manual for details and examples
        {
          //if it is last char, parsing's done
          // and the whole thing is valid.
          if (i == len - 1)
            isSwipeDataValid = true;
          //there's unexpected char after ending character \x0D
          else
            break;
        }
        //unexpected char when expecting a track
        else
          break;
      } else {
        //skip track content
        if (ps_isISO ? (b != 0x3F) : (b != 0x7F))
          continue;
        //track ended
        else {

          //save track
          /// tracks[ps_tracksI++] = [NSData dataWithBytes:bytes + ps_trackStart length: i+1-ps_trackStart];
          switch (ps_tracksI) {
          case 0: {
            for (int j = 0, k = ps_trackStart; j <= i + 1 - ps_trackStart; k++, j++) {
              track1[j] = bytes[k];
            }
            break;
          }
          case 1: {
            for (int j = 0, k = ps_trackStart; j <= i + 1 - ps_trackStart; k++, j++) {
              track2[j] = bytes[k];
            }
            break;
          }
          case 2: {
            for (int j = 0, k = ps_trackStart; j <= i + 1 - ps_trackStart; k++, j++) {
              track3[j] = bytes[k];
            }
            break;
          }
          }
          ps_tracksI++;
          ps_trackStart = i + 1;
          ps_isOutsideTrack = true;
        }
      }
    }

    //save result
    this.isValid = isSwipeDataValid;
    this.track1 = track1;
    this.track2 = track2;
    this.track3 = track3;
  }

} //class
